server/blob/blob-deduplicating/src/main/scala/GC.scala

```diff
@@ -0,0 +1,192 @@
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+package org.apache.james.server.blob.deduplication
+
+import org.apache.james.blob.api.BlobId
+
+sealed abstract class Generation extends Comparable[Generation] {
+  def previous: Generation
+  def previous(times: Long): Generation =
+    (0L until times).foldLeft(this)((generation, _) => generation.previous)
+
+  def next: Generation
+  def next(times: Long): Generation =
+    (0L until times).foldLeft(this)((generation, _) => generation.next)
+
+  def collectibles(targetedGeneration: Generation): Set[Generation] =
+    Generation.range(next, targetedGeneration.previous(GC.temporization)).toSet
+
+  def <(that : Generation): Boolean = compareTo(that) < 0
+  def <=(that : Generation): Boolean = compareTo(that) <= 0
+  def >(that : Generation): Boolean = compareTo(that) > 0
+  def >=(that : Generation): Boolean = compareTo(that) >= 0
+}
+
+case class ValidGeneration(id: Long) extends Generation {
+  override def previous: Generation = Generation(id - 1)
+
+  override def next: Generation = copy(id + 1)
+
+  override def compareTo(t: Generation): Int = t match {
+    case InvalidGeneration => 1
+    case that : ValidGeneration => id.compareTo(that.id)
+  }
+
+}
+
+case object InvalidGeneration extends Generation {
+  override def previous: Generation = InvalidGeneration
+
+  override def next: Generation = Generation.first
+
+  override def compareTo(t: Generation): Int = t match {
+    case InvalidGeneration => 0
+    case _ : ValidGeneration => -1
+  }
+}
+
+object Generation {
+  val first: Generation = apply(0)
+
+  def apply(id: Long): Generation = {
+    if (id < 0) {
+      InvalidGeneration
+    } else {
+      ValidGeneration(id)
+    }
+  }
+
+  def range(start : Generation, end : Generation) : Seq[Generation] = (start, end) match {
+    case (InvalidGeneration, InvalidGeneration) => Seq(InvalidGeneration)
+    case (ValidGeneration(_), InvalidGeneration) => Nil
+    case (InvalidGeneration, ValidGeneration(id)) =>  InvalidGeneration +: (0L to id).map(Generation.apply)
+    case (ValidGeneration(id1), ValidGeneration(id2)) => (id1 to id2).map(Generation.apply)
+  }
+}
+
+case class Iteration(id: Long, processedGenerations: Set[Generation], lastGeneration: Generation) {
+  def next(generations: Set[Generation], lastGeneration: Generation): Iteration = Iteration(id + 1, generations, lastGeneration)
+}
+
+object Iteration {
+  def initial: Iteration = Iteration(0, Set(InvalidGeneration), InvalidGeneration)
+}
+
+case class ExternalID(id: String) // TODO
```

Comment:
What's the purpose of this ExternalId ?

Reply:
Represent the source of the reference (ie. mail ids)